### PR TITLE
joystick: get HID top-level collection preparsed data directly from RawInput API

### PR DESCRIPTION
**Pull request by DJm00n**
…

### CODE REVIEW

File: src/joystick/windows/SDL_rawinputjoystick.c [hunk 1]
```diff
@@ -734,7 +739,7 @@ RAWINPUT_AddDevice(HANDLE hDevice)
         device->guid.data[15] = 0;
     }
 
-    hFile = CreateFileA(dev_name, GENERIC_READ, FILE_SHARE_READ|FILE_SHARE_WRITE, NULL, OPEN_EXISTING, 0, NULL);
+    hFile = CreateFileA(dev_name, 0, FILE_SHARE_READ | FILE_SHARE_WRITE, NULL, OPEN_EXISTING, 0, NULL);
```
Comment (cgutman) [inline src/joystick/windows/SDL_rawinputjoystick.c:742]: What's the reason for this change?

File: src/joystick/windows/SDL_rawinputjoystick.c [hunk 2]
```diff
@@ -714,6 +713,12 @@ RAWINPUT_AddDevice(HANDLE hDevice)
     device->is_xinput = SDL_TRUE;
     device->is_xboxone = GuessControllerType(device->vendor_id, device->product_id) == k_eControllerType_XBoxOneController;
 
+    /* Get HID Top-Level Collection Preparsed Data */
+    size = 0;
+    CHECK(GetRawInputDeviceInfoA(hDevice, RIDI_PREPARSEDDATA, NULL, &size) != (UINT)-1);
+    CHECK(device->preparsed_data = (PHIDP_PREPARSED_DATA)SDL_calloc(size, sizeof(BYTE)));
+    CHECK(GetRawInputDeviceInfoA(hDevice, RIDI_PREPARSEDDATA, device->preparsed_data, &size) != (UINT)-1);
```
Comment (cgutman) [inline src/joystick/windows/SDL_rawinputjoystick.c:720]: What is the advantage of this over `HidD_GetPreparsedData()`?

File: src/joystick/windows/SDL_rawinputjoystick.c (hunk 1 above)
Comment (DJm00n) [reply src/joystick/windows/SDL_rawinputjoystick.c:742]: We don’t need read access to read device info from handle. So just don’t request it.

File: src/joystick/windows/SDL_rawinputjoystick.c (hunk 2 above)
Comment (DJm00n) [reply src/joystick/windows/SDL_rawinputjoystick.c:720]: RawInput API provides pp data and I think it is clearer to request it like that. Just for consistency.

File: src/joystick/windows/SDL_rawinputjoystick.c (hunk 1 above)
Comment (cgutman) [reply src/joystick/windows/SDL_rawinputjoystick.c:742]: Yep, that looks accurate. The underlying HID IOCTLs are all declared as FILE_ANY_ACCESS.

**Review by cgutman — APPROVED** [review]

File: src/joystick/windows/SDL_rawinputjoystick.c (hunk 1 above)
Comment (slouken) [inline src/joystick/windows/SDL_rawinputjoystick.c:742]: We ran into driver bugs when specifying 0 instead of GENERIC_READ. Is this an important part of this patch?

Comment (DJm00n) [reply src/joystick/windows/SDL_rawinputjoystick.c:742]: no, its not important. but I have added it be on par with hidapi code.

Comment (slouken) [reply src/joystick/windows/SDL_rawinputjoystick.c:742]: Please remove it from this PR.

Comment (DJm00n) [reply src/joystick/windows/SDL_rawinputjoystick.c]: done
